### DISCUSSION TITLE
Lesson 3: Fund Me Higher Level Discussion

### DISCUSSION THREAD

**Discussion by JohnEMagee** [smartcontractkit/full-blockchain-solidity-course-py#1186] [Q&A] [answered]
I just finished the first part of the FundMe Contract and while I follow along the code I'm a bit unclear on the math.  (Solidity not using decimals is one of many things I find difficult to grasp coming from other languages). There's a lot of multiplying and dividing that to me don't seem well explained.  Is everything in Solidity expressed in 'WEI' bits.  Why is the USD value of ETH expressed in so many digits and am I supposed to know to divide by 8 to get there somehow?  I'm enjoying the tutorials and the coding but this lack of clarity to me on the size of numbers and some sort of default value (and best how to deal with it) is a bit frustrating.  I mean I get the 'using the calculator' to verify but how do you know?  To me it's a pretty important concept I have to get before moving forward.

**Comment by Noveleader** [comment]
When  you make function fund() we set minimumUSD to 50$ and multiply it with 10^18, hence converting it according to wei. This is because when we use getConversionRate we also enter the value in wei and hence we get ethAmountinUSD in terms of wei according to the eth price in dollars and we compare it with 50$ using require.

TLDR: In getConversionRate function we enter the value in wei and hence in return we get out dollar value multiplied with 10^18 (once put something in wei in getConversionRate) hence comparing it 50*10**18

**Comment by JohnEMagee** [comment]
Are you saying input matches output?  If in get conversion rate I entered 1ETH instead of 1WEI it would have been in ETH?
Going backwards getPrice returns the answer multiplied by 10 ** 10.  So I guess my question is when you run priceFeed.latestRoundData() (which I just did) and get 2956508568200000000000 - that's in WEI, but if we removed all the zeroes, is that in GWEI?  (as an aside, nanoETH and something similar might have made more sense I think).  

THe USD price would be 2956.56056820 - with decimals, without decimals it's obviously a larger number but how do we know we have to divide by 10**8 to get that value...I sort of get the functionality but I don't understand the base logic between the return (the whole no decimcals thing in solidity seems overly complicated, is there a logic behind that?)

**Reply by Noveleader** [reply]
Versions now don't deal with decimals but in a future update, we may get that for now it only deals with the whole numbers.

**Comment by cromewar** [comment]
Hello @JohnEMagee well this is a tricky question, let's consider the following:

1. the price feed address returns a number with different decimals depending on the chain and the currency, is recommended to check the chainlink docs.
2. Is not an obligation to use 18 decimals, but is a good practice as WEI has 18 is easier to manage everything on WEI.
3. Instead of using a lot o zeros you can always use `* 10 ** 18` .

**Comment by n4n0b1t3** [comment, marked as answer]
Hi @JohnEMagee , I recently wrote [a lengthy post exactly about this topic](https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/1153). Please let me know if this was helpful. https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/1153

**Reply by JohnEMagee** [reply]
I will read the article in depth but I already found the issue and the issue is this was not made clear in the video - which should be correct

_When you copy the data feed address, you will find the information of DEC in the column next to it, for this data feed a DEC value of 8 is displayed_

That simple piece of information conveyed clearly in the video would have saved me a lot of consternation (the rest of the consternation is on whoever created solidity and said 'nah, we don't need decimal points, not in a coding language' - that seems odd)

**Reply by n4n0b1t3** [reply]
I am using the decimal value programmatically to create a USD-WEI equivalent. It can be obtained with `priceFeed.decimals()`

```python
	function getEntranceFee() public view returns (uint256 _entranceFeeInWei) {
		(int256 exchangeRate, uint8 decimals) = getExchangeRate();
		uint256 ticketPriceUsdWeiEquivalent = TICKET_PRICE_USD * 10**18;
		uint256 exchangeRateWeiEquivalent = uint256(exchangeRate) *	10**(18 - decimals);
		uint256 entranceFeeInWei = (ticketPriceUsdWeiEquivalent * 10**18) /	exchangeRateWeiEquivalent;
		return entranceFeeInWei;
	}


	function getExchangeRate() public view returns (int256 exchangeRate, uint8 decimals){
		(, int256 _exchangeRate, , , ) = priceFeed.latestRoundData();
		return (_exchangeRate, priceFeed.decimals());
	}
```

More details in the aforementioned article.

**Comment by JohnEMagee** [comment]
Thanks for everyones help the math part of it makes sense to me now but I realize there's a piece I was missing deeper that maybe I missed in the lecture.  If i enter (currently) 10000 WEI the fund will reject it as too low,.  If I enter 10000 GWEI, it gets rejected as too low.  However, if I enter .03 ETH it will accept it.  This is the variable _msg.value_.  But for this to work msg.value must have a 'default' currency.  Is _msg.value_ automatically converted to a standard value (i'm assuming WEI/GWEI probably).  That might be the key piece I was missing, if the _msg.value_ is always standardized , (logically it seems WEI or GWEI since you can't do .03 ETH as a value) then it would make sense to me.  I hope I'm right about this...and if I am right, is this covered in the solidity docs and I just haven't read them yet?

**Reply by n4n0b1t3** [reply]
msg.value is always in WEI, as everything is in solidity.

**Reply by n4n0b1t3** [reply]
When I wrote the article I noticed, that it is not so easy to coax python to print out numbers as decimals and not e-notation. You would need that to paste it into a [wei-eth converter](https://eth-converter.com/) for example.
Here is a code that would print out the decimal numbers with all the zeros.
```python
>>> from decimal import Decimal
>>> value = Decimal("247223226886") * Decimal(10 ** 10)   
>>> print(value)
2472232268860000000000
```

**Reply by cromewar** [reply]
Nice approach!
